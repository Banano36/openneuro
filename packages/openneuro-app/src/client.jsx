/**
 * Browser client entrypoint - see server.tsx for SSR entrypoint
 */
import './scripts/utils/global-polyfill'
import './scripts/apm.js'
import { ApolloProvider, InMemoryCache } from '@apollo/client'
import { createClient } from '@openneuro/client'
import React from 'react'
import ReactDOM from 'react-dom'
import { BrowserRouter, Route } from 'react-router-dom'
import App from './scripts/app'
import Index from './scripts/index'
import analyticsWrapper from './scripts/utils/analytics'
import { version } from './lerna.json'
import { config } from './scripts/config'
import * as gtag from './scripts/utils/gtag'
import { relayStylePagination } from '@apollo/client/utilities'
import redesignStyles from '@openneuro/components/page/page.scss'

gtag.initialize(config.analytics.trackingIds)

ReactDOM.render(
  <App>
    <ApolloProvider
      client={createClient(`${config.url}/crn/graphql`, {
        clientVersion: version,
        enableWebsocket: false,
        cache: new InMemoryCache({
          typePolicies: {
            Query: {
              fields: {
                advancedSearch: relayStylePagination(),
              },
            },
          },
        }),
      })}>
      <BrowserRouter>
        <Route component={analyticsWrapper(Index)} />
      </BrowserRouter>
    </ApolloProvider>
  </App>,
  document.getElementById('main'),
)
