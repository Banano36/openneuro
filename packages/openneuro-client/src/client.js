import 'cross-fetch/polyfill'
import { ApolloClient } from 'apollo-client'
import { setContext } from 'apollo-link-context'
import { InMemoryCache } from 'apollo-cache-inmemory'
import { createUploadLink } from 'apollo-upload-client'
import { ApolloLink, split, Observable } from 'apollo-link'
import { WebSocketLink } from 'apollo-link-ws'
import { getMainDefinition } from 'apollo-utilities'
import FormData from 'form-data'
import semver from 'semver'
import * as files from './files'
import * as datasets from './datasets'
import * as snapshots from './snapshots'
import * as users from './users'

const cache = new InMemoryCache({
  freezeResults: true,
})

/**
 * Setup a client for working with the OpenNeuro API
 *
 * @param {string} uri
 */
const createClient = (uri, getAuthorization, fetch, clientVersion) => {
  const link = createLink(uri, getAuthorization, fetch, clientVersion)
  return new ApolloClient({ uri, link, cache, connectToDevTools: true })
}

const authLink = getAuthorization =>
  setContext((_, { headers }) => {
    // Passthrough any headers but add in authorization if set
    const token = getAuthorization()
    let cookie = {}
    if (token) {
      const tokenString = `accessToken=${token}`
      cookie = {
        cookie: tokenString,
      }
    }
    return {
      headers: Object.assign(cookie, headers),
    }
  })

const wsLink = uri => {
  const root = uri.replace('http', 'ws').replace('/crn', '')
  const subscriptions = '-subscriptions'
  const link = root + subscriptions
  return new WebSocketLink({
    uri: link,
    options: {
      reconnect: true,
    },
  })
}

const middlewareAuthLink = (uri, getAuthorization, fetch) => {
  // We have to setup authLink to inject credentials here
  const httpUploadLink = createUploadLink({
    uri,
    fetch: fetch === null ? undefined : fetch,
    serverFormData: FormData,
    credentials: 'same-origin',
  })
  if (getAuthorization) {
    return authLink(getAuthorization).concat(httpUploadLink)
  } else {
    return httpUploadLink
  }
}

const parse = version => [semver.major(version), semver.minor(version)]
const checkVersions = (serverVersion, clientVersion) => {
  if ([serverVersion, clientVersion].every(semver.valid)) {
    const [serverMajor, serverMinor] = parse(serverVersion)
    const [clientMajor, clientMinor] = parse(clientVersion)
    if (serverMajor > clientMajor || serverMinor > clientMinor) {
      console.warn(
        `Your openNeuro client is out of date (v${clientVersion}). We strongly recommend you update to the latest version (v${serverVersion}) for an optimal experience.`,
      )
    } else if (
      serverMajor < clientMajor ||
      (serverMajor === clientMajor && serverMinor < clientMinor)
    ) {
      // panic, then
      console.warn(
        'Your openNeuro client is out of date. We strongly recommend you update to the most recent version for an optimal experience.',
      )
    }
  }
}

const compareVersionsLink = clientVersion =>
  new ApolloLink(
    (operation, forward) =>
      new Observable(observer =>
        forward(operation).subscribe({
          next: result => {
            const serverVersion = result.extensions.openneuro.version
            // alert user if major/minor versions are not in sync
            checkVersions(serverVersion, clientVersion)
            observer.next(result)
          },
          error: console.error,
          complete: observer.complete.bind(observer),
        }),
      ),
  )

const createLink = (uri, getAuthorization, fetch, clientVersion) => {
  // We have to setup authLink to inject credentials here

  // server-side link
  let link = middlewareAuthLink(uri, getAuthorization, fetch)
  let ws

  try {
    // browser-side link
    ws = process.browser ? wsLink(uri) : null
  } catch (e) {
    // Don't die when websocket setup fails
  }
  if (ws) {
    link = split(
      ({ query }) => {
        const { kind, operation } = getMainDefinition(query)
        return kind === 'OperationDefinition' && operation === 'subscription'
      },
      ws,
      middlewareAuthLink(uri, getAuthorization, fetch),
    )
  }

  return ApolloLink.from([compareVersionsLink(clientVersion), link])
}

export { files, datasets, snapshots, users }
export default createClient
