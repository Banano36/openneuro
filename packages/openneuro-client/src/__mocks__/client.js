import { ApolloClient, InMemoryCache } from '@apollo/client'
import { SchemaLink } from '@apollo/client/link/schema'
import { addMocksToSchema } from '@graphql-tools/mock'
import { makeExecutableSchema } from '@graphql-tools/schema'
import { typeDefs } from '@openneuro/server/src/graphql/schema'
import resolvers from '@openneuro/server/src/graphql/resolvers/index.js'

/**
 * Generate sequential dataset ids for tests
 * @param {number} n Starting string value
 */
function* idGenerator(n = 0) {
  while (true) {
    const idStr = String(++n).padStart(6, '0')
    yield `ds${idStr}`
  }
}

export const datasetIds = idGenerator()
export const testDsId = datasetIds.next().value
export const testTime = new Date()

// Workaround to restore this test, it should be refactored with openneuro-client updates
const schema = makeExecutableSchema({
  typeDefs,
  resolvers,
})

addMocksToSchema({
  schema,
  mocks: {
    Dataset: (root, { id }) => ({
      id: () => (id ? id : datasetIds.next().value),
      created: testTime,
      modified: testTime,
    }),
    DateTime: () => {
      return testTime
    },
    BigInt: () => {
      return 8589934592
    },
  },
})

export const createClient = uri => {
  const cache = new InMemoryCache()
  const link = new SchemaLink({
    schema,
    context: {
      user: '1234',
      userInfo: {
        id: '1234',
      },
    },
  })
  // @ts-expect-error
  return new ApolloClient({ uri, link, cache })
}
