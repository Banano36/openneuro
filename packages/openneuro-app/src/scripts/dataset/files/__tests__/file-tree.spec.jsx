import React from "react"
import { fireEvent, render, screen } from "@testing-library/react"
import { MockedProvider } from "@apollo/client/testing"
import FileTree, { fileTreeLevels, unescapePath } from "../file-tree"

// official Jest workaround for mocking methods not implemented in JSDOM
window.matchMedia = window.matchMedia ||
  function () {
    return {
      matches: false,
      addListener: vi.fn(),
      removeListener: vi.fn(),
    }
  }

const datasetFiles = [
  {
    id: "b42624c6aea63fc5e3f6f3e712d9e44adc4dbfdc",
    key: "834b88a80109f1b38e0ab85999090170889469ce",
    filename: "CHANGES",
    size: 59,
    directory: false,
    annexed: false,
  },
  {
    id: "63888a199a5ce37377b1cd708cda59577dad218f",
    key: "fa84e5f958ec72d42b3e196e592f6db9f7104b19",
    filename: "README",
    size: 709,
    directory: false,
    annexed: false,
  },
  {
    id: "a2922b427c5c997e77ce058e0ce57ffd17123a7f",
    key: "0b1856b91c11f67098ce60114417a62dd55730a5",
    filename: "dataset_description.json",
    size: 172,
    directory: false,
    annexed: false,
  },
  {
    id: "a6378eea201d9cad639e0bee328e03132d30489d",
    key: null,
    filename: "sub-01",
    size: 0,
    directory: true,
    annexed: false,
  },
]

const anatDir = {
  id: "9703f3f6b4137c2b86a3a712eb34d78bdec9fd72",
  key: null,
  filename: "sub-01:anat",
  size: 0,
  directory: true,
  annexed: false,
}

const niftiFile = {
  id: "69fd3617b27125c433ea5f8e0e2052c31828c253",
  key: "MD5E-s311112--bc8bbbacfd2ff823c2047ead1afec9b3.nii.gz",
  filename: "sub-01:anat:sub-01_T1w.nii.gz",
  size: 311112,
  directory: false,
  annexed: true,
}

describe("fileTreeLevels()", () => {
  it("handles top level files correctly", () => {
    const { childFiles, currentFiles } = fileTreeLevels("", datasetFiles)
    expect(currentFiles).toEqual(
      expect.arrayContaining([
        currentFiles.find(
          (f) => f.id === "a2922b427c5c997e77ce058e0ce57ffd17123a7f",
        ),
      ]),
    )
    expect(childFiles).toEqual({})
  })
  it("passes childFiles to the next level", () => {
    const levelOneFiles = [...datasetFiles, anatDir]
    const { childFiles, currentFiles } = fileTreeLevels("", levelOneFiles)
    expect(currentFiles).toEqual(
      expect.arrayContaining([
        currentFiles.find(
          (f) => f.id === "a2922b427c5c997e77ce058e0ce57ffd17123a7f",
        ),
      ]),
    )
    expect(childFiles["sub-01"]).toEqual(expect.arrayContaining([anatDir]))
  })
  it("passes two level deep childFiles to the next level", () => {
    const levelTwoFiles = [...datasetFiles, anatDir, niftiFile]
    const { childFiles, currentFiles } = fileTreeLevels("", levelTwoFiles)
    expect(currentFiles).toEqual(
      expect.arrayContaining([
        currentFiles.find(
          (f) => f.id === "a2922b427c5c997e77ce058e0ce57ffd17123a7f",
        ),
      ]),
    )
    expect(childFiles["sub-01"]).toEqual(expect.arrayContaining([niftiFile]))
  })
  it("passes two level deep childFiles to the next level", () => {
    const levelTwoFiles = [anatDir, niftiFile]
    const { childFiles, currentFiles } = fileTreeLevels("sub-01", levelTwoFiles)
    expect(currentFiles).toEqual(
      expect.arrayContaining([
        currentFiles.find(
          (f) => f.id === "9703f3f6b4137c2b86a3a712eb34d78bdec9fd72",
        ),
      ]),
    )
    expect(childFiles["sub-01:anat"]).toEqual(
      expect.arrayContaining([niftiFile]),
    )
  })
})

describe("FileTree component", () => {
  it("renders with default props", () => {
    const { asFragment } = render(<FileTree />)
    expect(asFragment()).toMatchSnapshot()
  })
  it("expands and closes when clicked", () => {
    render(
      <MockedProvider>
        <FileTree name="Top Level" />
      </MockedProvider>,
    )
    // Test the folder icon is closed
    expect(screen.getByLabelText("Top Level").firstChild).toHaveClass(
      "fa-folder",
    )
    expect(screen.getByLabelText("Top Level").firstChild).not.toHaveClass(
      "fa-folder-open",
    )
    // Click it
    fireEvent.click(screen.getByLabelText("Top Level"))
    // Test that it is now open
    expect(screen.getByLabelText("Top Level").firstChild).toHaveClass(
      "fa-folder-open",
    )
    expect(screen.getByLabelText("Top Level").firstChild).not.toHaveClass(
      "fa-folder",
    )
  })
  describe("unescapePath()", () => {
    it("does not alter an already escaped path", () => {
      expect(unescapePath("sub-01/anat")).toBe("sub-01/anat")
    })
    it("does unescapes any : characters", () => {
      expect(unescapePath("sub-01:anat")).toBe("sub-01/anat")
    })
    it("unescapes multiple : characters", () => {
      expect(unescapePath("sub-01:anat:image.nii.gz")).toBe(
        "sub-01/anat/image.nii.gz",
      )
    })
  })
})
