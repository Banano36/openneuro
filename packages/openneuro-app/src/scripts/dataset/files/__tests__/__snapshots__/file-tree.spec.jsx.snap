// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileTree component renders with default props 1`] = `
<DocumentFragment>
  <article
    class="file-tree accordion "
  >
    <div
      class="accordion-title "
    >
      <span
        aria-label=""
        class="file-icon on-icon  "
        role="img"
      >
        <i
          class="fas fa-folder"
        />
      </span>
    </div>
    <div
      class="accordion-item  collapsed "
    >
      <div
        class="accordion-content"
      >
        <ul
          class="child-files"
        />
      </div>
    </div>
  </article>
</DocumentFragment>
`;
